File: BM.BLL/BorrowManager.cs
using System;
using System.Data;
using BookManagement.DAL;
using BookManagement.Model;

namespace BookManagement.BLL
{
    public class BorrowManager
    {
        private BorrowService borrowService = new BorrowService();
        private ReaderService readerService = new ReaderService();
        private BookService bookService = new BookService();

        /// <summary>
        /// 验证借阅条件
        /// </summary>
        public (bool canBorrow, string message) ValidateBorrow(int readerId, int bookId)
        {
            try
            {
                // 1. 检查读者是否存在
                DataTable readerDt = readerService.GetReaderById(readerId);
                if (readerDt == null || readerDt.Rows.Count == 0)
                    return (false, "读者不存在！");

                // 2. 检查读者状态
                int readerStatus = Convert.ToInt32(readerDt.Rows[0]["Status"]);
                if (readerStatus != 0) // 0:正常
                    return (false, "读者状态异常，无法借书！");

                // 3. 检查图书是否存在
                DataTable bookDt = bookService.GetBookById(bookId);
                if (bookDt == null || bookDt.Rows.Count == 0)
                    return (false, "图书不存在！");

                // 4. 检查图书库存
                int availableCount = Convert.ToInt32(bookDt.Rows[0]["AvailableCount"]);
                if (availableCount <= 0)
                    return (false, "该图书已全部借出！");

                // 5. 检查读者借阅上限
                int categoryId = Convert.ToInt32(readerDt.Rows[0]["ReaderCategoryId"]);
                DataTable categoryDt = readerService.GetReaderCategoryById(categoryId);
                if (categoryDt != null && categoryDt.Rows.Count > 0)
                {
                    int maxBorrow = Convert.ToInt32(categoryDt.Rows[0]["MaxBorrowCount"]);

                    // 获取读者当前借阅数量
                    DataTable borrowDt = borrowService.GetReaderBorrowInfo(readerId);
                    int currentBorrowCount = borrowDt?.Rows.Count ?? 0;

                    if (currentBorrowCount >= maxBorrow)
                        return (false, $"已达到最大借阅数量（{maxBorrow}本）！");
                }

                return (true, "可以借阅");
            }
            catch (Exception ex)
            {
                return (false, $"验证失败：{ex.Message}");
            }
        }

        /// <summary>
        /// 执行借书操作
        /// </summary>
        public bool BorrowBook(int readerId, int bookId, DateTime borrowDate, DateTime dueDate)
        {
            // 先验证
            var validation = ValidateBorrow(readerId, bookId);
            if (!validation.canBorrow)
                throw new Exception(validation.message);

            // 执行借书
            return borrowService.BorrowBook(readerId, bookId, borrowDate, dueDate);
        }

        /// <summary>
        /// 获取读者借阅信息
        /// </summary>
        public DataTable GetReaderBorrowInfo(int readerId)
        {
            return borrowService.GetReaderBorrowInfo(readerId);
        }
        /// <summary>
        /// 查询借阅记录
        /// </summary>
        public DataTable SearchBorrowRecords(int? readerId, string isbn, int? borrowId)
        {
            return borrowService.SearchBorrowRecords(readerId, isbn, borrowId);
        }
        /// <summary>
        /// 多条件查询所有借阅记录（支持分页）
        /// </summary>
        /// <summary>
        /// 多条件查询所有借阅记录(支持分页)
        /// </summary>
        public DataTable SearchAllBorrowRecords(int? readerId, string readerName, string bookName,
            string isbn, int? status, string category,
            DateTime? borrowFrom, DateTime? borrowTo,
            DateTime? returnFrom, DateTime? returnTo,
            int pageIndex, int pageSize, out int totalRecords)
        {
            return borrowService.SearchAllBorrowRecords(
                readerId, readerName, bookName, isbn,
                status, category, borrowFrom, borrowTo,
                returnFrom, returnTo, pageIndex, pageSize,
                out totalRecords
            );
        }


        /// <summary>
        /// 还书操作
        /// </summary>
        public decimal CalculateLateFee(int borrowId, DateTime actualReturnDate)
        {
            try
            {
                // 获取借阅记录详情
                DataTable borrowDt = borrowService.GetBorrowRecordById(borrowId);
                if (borrowDt == null || borrowDt.Rows.Count == 0)
                    throw new Exception("借阅记录不存在");

                // 获取应还日期
                DateTime dueDate = Convert.ToDateTime(borrowDt.Rows[0]["DueDate"]);

                // 计算逾期天数
                int overdueDays = 0;
                if (actualReturnDate > dueDate)
                {
                    overdueDays = (actualReturnDate - dueDate).Days;
                }

                // 获取读者类别ID
                int readerCategoryId = Convert.ToInt32(borrowDt.Rows[0]["ReaderCategoryId"]);

                // 获取读者类别的逾期费率
                DataTable categoryDt = readerService.GetReaderCategoryById(readerCategoryId);
                if (categoryDt == null || categoryDt.Rows.Count == 0)
                    throw new Exception("读者类别不存在");

                decimal lateFeePerDay = Convert.ToDecimal(categoryDt.Rows[0]["LateFeePerDay"]);

                return overdueDays * lateFeePerDay;
            }
            catch (Exception ex)
            {
                throw new Exception($"计算逾期费用失败：{ex.Message}");
            }
        }

        public (bool success, string message, decimal lateFee) ReturnBook(int borrowId, DateTime returnDate, decimal lateFee, string remark)
        {
            try
            {
                // 计算逾期费用
                decimal calculatedLateFee = CalculateLateFee(borrowId, returnDate);

                // 如果传入的逾期费用为0，使用计算值
                if (lateFee == 0 && calculatedLateFee > 0)
                {
                    lateFee = calculatedLateFee;
                }

                // 执行还书
                bool result = borrowService.ReturnBook(borrowId, returnDate, lateFee, remark);

                if (result)
                {
                    return (true, "还书成功", lateFee);
                }
                else
                {
                    return (false, "还书失败", 0);
                }
            }
            catch (Exception ex)
            {
                return (false, $"还书失败：{ex.Message}", 0);
            }
        }
        public DataTable GetBorrowRecordById(int borrowId)
        {
            return borrowService.GetBorrowRecordById(borrowId);
        }

    }
}

File: Python/bll/borrow_manager.py
from dal.borrow_service import BorrowService
from dal.reader_service import ReaderService
from dal.book_service import BookService
from datetime import datetime


class BorrowManager:
    def __init__(self):
        self.borrow_service = BorrowService()
        self.reader_service = ReaderService()
        self.book_service = BookService()

    def validate_borrow(self, reader_id: int, book_id: int) -> tuple:
        try:
            reader_rows = self.reader_service.get_reader_by_id(reader_id)
            if not reader_rows:
                return False, "读者不存在！"

            reader_status = reader_rows[0]['Status']
            if reader_status != 0:
                return False, "读者状态异常，无法借书！"

            book_rows = self.book_service.get_book_by_id(book_id)
            if not book_rows:
                return False, "图书不存在！"

            available_count = book_rows[0]['AvailableCount']
            if available_count <= 0:
                return False, "该图书已全部借出！"

            category_id = reader_rows[0]['ReaderCategoryId']
            category_rows = self.reader_service.get_reader_category_by_id(category_id)
            if category_rows:
                max_borrow = category_rows[0]['MaxBorrowCount']
                borrow_rows = self.borrow_service.get_reader_borrow_info(reader_id)
                current_borrow_count = len(borrow_rows) if borrow_rows else 0
                if current_borrow_count >= max_borrow:
                    return False, f"已达到最大借阅数量（{max_borrow}本）！"

            return True, "可以借阅"
        except Exception as ex:
            return False, f"验证失败：{str(ex)}"

    def borrow_book(self, reader_id: int, book_id: int, borrow_date: str, due_date: str) -> bool:
        can_borrow, message = self.validate_borrow(reader_id, book_id)
        if not can_borrow:
            raise Exception(message)
        return self.borrow_service.borrow_book(reader_id, book_id, borrow_date, due_date)

    def get_reader_borrow_info(self, reader_id: int):
        return self.borrow_service.get_reader_borrow_info(reader_id)

    def search_borrow_records(self, reader_id: int = None, isbn: str = None, borrow_id: int = None, status: int = None):
        return self.borrow_service.search_borrow_records(reader_id, isbn, borrow_id, status)

    def search_all_borrow_records(self, reader_id: int = None, reader_name: str = None,
                                  book_name: str = None, isbn: str = None,
                                  status: int = None, borrow_from: str = None,
                                  borrow_to: str = None, return_from: str = None,
                                  return_to: str = None, page_index: int = 1,
                                  page_size: int = 10) -> tuple:
        return self.borrow_service.search_all_borrow_records(
            reader_id, reader_name, book_name, isbn, status,
            borrow_from, borrow_to, return_from, return_to, page_index, page_size
        )

    def calculate_late_fee(self, borrow_id: int, actual_return_date: str) -> float:
        try:
            borrow_rows = self.borrow_service.get_borrow_record_by_id(borrow_id)
            if not borrow_rows:
                raise Exception("借阅记录不存在")

            due_date_str = borrow_rows[0]['DueDate']
            due_date = datetime.strptime(due_date_str, '%Y-%m-%d')
            actual_date = datetime.strptime(actual_return_date, '%Y-%m-%d')

            overdue_days = 0
            if actual_date > due_date:
                overdue_days = (actual_date - due_date).days

            reader_category_id = borrow_rows[0]['ReaderCategoryId']
            category_rows = self.reader_service.get_reader_category_by_id(reader_category_id)
            if not category_rows:
                raise Exception("读者类别不存在")

            late_fee_per_day = category_rows[0]['LateFeePerDay']
            return overdue_days * late_fee_per_day
        except Exception as ex:
            raise Exception(f"计算逾期费用失败：{str(ex)}")

    def return_book(self, borrow_id: int, return_date: str, late_fee: float, remark: str) -> tuple:
        try:
            calculated_late_fee = self.calculate_late_fee(borrow_id, return_date)
            if late_fee == 0 and calculated_late_fee > 0:
                late_fee = calculated_late_fee

            result = self.borrow_service.return_book(borrow_id, return_date, late_fee, remark)
            if result:
                return True, "还书成功", late_fee
            else:
                return False, "还书失败", 0
        except Exception as ex:
            return False, f"还书失败：{str(ex)}", 0

    def get_borrow_record_by_id(self, borrow_id: int):
        return self.borrow_service.get_borrow_record_by_id(borrow_id)